docs/writing-rules/glossary.md

```diff
@@ -0,0 +1,124 @@
+---
+slug: glossary
+title: SAST and rule-writing glossary
+hide_title: true
+description: Definitions of static analysis and Semgrep rule-writing terms.
+tags:
+  - Rule writing
+---
+
+# Static analysis and rule-writing glossary
+
+The definitions provided here are specific to each term's meaning and use in Semgrep.
+
+## Constant propagation
+
+Refers to state of a variable remaining constant throughout the program. Semgrep can analyze whether a variable carries a constant value at a given point. Both Semgrep OSS and Semgrep Pro Engine perform this analysis, with Semgrep Pro able to track the propagation across files.
```

Comment:
Hi @emjin can you explain what the difference is between:

A: Constant propagation can analyze whether a variable carries a constant value at a given point.
B: Constant propagation is a type of analysis where values known to be constant are substituted in later uses. 

what does it mean to be "known as constant"?

In the docs, it says:
> This analysis tracks whether a variable must carry a constant value at a given point in the program. 

I was assuming that Semgrep is tracking whether some variable assignment such as X = 59 is always true?

Reply:
> By substitution, you mean that Semgrep can detect that print(x) is equal to print(2) (In your example below) because Semgrep can detect the assignment and propagate it. Semgrep "substitutes" the value of x and thus detects print(2)

I agree with this statement. However, I want to clarify that we only do this substitution for values we detect to be constant. That is so that we can avoid errors in our analysis.

> Constant here doesn't mean that the value never changes over the course of the program; it just means that we defined it as a constant in our rule, i.e. the string "abcd" is always "abcd" - because we defined it in the pattern operator.

I disagree with this statement. Constant does mean that the value never changes over the course of the program. We only propagate values for variables that are constant.